Fields default to Blank String

MichaelRomani opened the pull request:
- Summary
Closes issue #977 
Widget values default to blank string when value is null.

- Test plan
Values for title, image and body within redux store get set to '' when value is null (date currently being set to current date)

- Description for the changelog
Default values directly provided by each widget.

verythorough commented:
Deploy preview for *cms-demo* ready!

Built with commit ba1a709b5b4a2708ab11465886e2f35c7158e5d6

https://deploy-preview-1126--cms-demo.netlify.com

verythorough commented:
Deploy preview for *netlify-cms-www* ready!

Built with commit ba1a709b5b4a2708ab11465886e2f35c7158e5d6

https://deploy-preview-1126--netlify-cms-www.netlify.com

erquhart commented:
This approach works, but I'd like to see a less involved method for handling defaults since every widget will need to tackle this concern.

Right now, we're intentionally setting empty fields to `null`: https://github.com/netlify/netlify-cms/blob/b4b584682473924556a41cae10f64f085b0f432b/src/actions/entries.js#L261-L263

A `null` value "represents the intentional absence of any object value", while `undefined` represents a value that simply hasn't been set. The latter is a more accurate description of fields that have no default value ([spec](http://www.ecma-international.org/ecma-262/5.1/#sec-4.3.9)), so I propose we remove the default setting of `null`, allowing empty fields to return `undefined` instead.

In doing so, we gain the ability to use React's [`defaultProps`](https://reactjs.org/docs/typechecking-with-proptypes.html#default-prop-values) to set empty values in widgets (they only work with `undefined` values).

Thoughts?

MichaelRomani commented:
Shawn, just to clarify, you'd like me to set a default for 'value' using defaultProps for each of the editor widgets that has 'value' within its propTypes object.  Is this correct?  Also, should I default each to an empty string including objects that might ultimately be set to a date / boolean?

erquhart commented:
Each value will either be a number, string, boolean, array, object, or an array/object containing these types. Dates, for example, are stringified by their widget on change. Most widgets will use an empty string, with the exception of the boolean, object, and list widget. Perhaps there are others, those are the only ones that come to mind at the moment.

MichaelRomani commented:
Thanks, I will make the adjustments and submit early this week!
-Mike

MichaelRomani commented:
I've updated the PR to include 'defaultProps' for values located within the editor widget files.  I did not set the default for the date/time as the current set up will only set the current date / time if value is not equal to an empty string.  I also updated the createEmptyDraft function to set all values as undefined in place of null.

tech4him1 commented on src/actions/entries.js:
```diff
@@ -259,7 +259,7 @@ export function createEmptyDraft(collection) {
     const dataFields = {};
     collection.get('fields', List()).forEach((field) => {
-      dataFields[field.get('name')] = field.get('default', null);
+      dataFields[field.get('name')] = undefined;
```
@MichaelRomani We still want to allow users to manually set a default here, just default it to `undefined` instead of `null`. The code would be `field.get('default')`, just removing `null` as the second param like @erquhart mentioned.

MichaelRomani commented:
@tech4him1  - I've made the adjustment, setting it equal to - field.get('default').

tech4him1 commented on src/components/EditorWidgets/Markdown/MarkdownPreview/index.js:
```diff
@@ -15,4 +15,8 @@ MarkdownPreview.propTypes = {
 };
 
+MarkdownPreview.defaultProps = {
+  value: '',
```
Is there a reason we need to set this on the MD preview, and not any of the other previews?

MichaelRomani replied:
Are you referring to the 'RawEditor' and 'VisualEditor' files?

MichaelRomani replied:
As far as the kitchen sink goes, I'll take a look this evening or tomorrow afternoon.  My apologies, I did not test creating a new sink.

tech4him1 replied:
No, I just noticed that you set a default value on the MarkdownPreview as well as the MarkdownControl.

erquhart commented:
Looking good @MichaelRomani!

The ultimate test is creating a new Kitchen Sink entry in the demo - currently failing: https://deploy-preview-1126--cms-demo.netlify.com/#/collections/kitchenSink

tech4him1 commented:
@MichaelRomani Maybe I missed it, but is there a reason a default isn't set for the ListControl?

tech4him1 commented:
@erquhart Is changing the default from `null` to `undefined` going to be a breaking change?

tech4him1 commented:
oops 😁

erquhart commented:
It would only be breaking for an installation if someone were intentionally checking for that null value in an external custom widget, in which case, the null value isn't a part of our published API, so it's not technically a breaking change. I also think the likelihood of real world problems resulting is very low, but I'm quite open to being told I'm wrong 😄.

MichaelRomani commented:
ListControl - A defaultProp for value was already set in this file as:
`value: List()` 

MarkdownPreview - I removed the defaultProps that had been added to this file.  I mistakenly thought it was receiving 'undefined' and needed to be set to ''.

Kitchen Sink Error -The error seems to have been coming from ObjectControl.  I think the issue is that the defaultProp should have been set to a Map Object as:
`Map()`.  
Locally, Kitchen Sink is now working.

Let me know if this makes sense and I'll push these changes up to this PR for review.

erquhart commented:
Sounds good, push it up!

tech4him1 commented:
@MichaelRomani Great, thanks!

erquhart commented:
Any thoughts from anyone about the breaking change [concern](https://github.com/netlify/netlify-cms/pull/1126#issuecomment-369085003)?

cc/ @tech4him1

tech4him1 commented:
@erquhart On netlify/netlify-cms#1126, does the CMS internally treat null differently than undefined? Maybe by writing a key as empty vs not writing at all, or anything like that?

If not, I don't see a big concern with backwards compatibility.

erquhart commented:
Not that I'm aware of.
